R/sensitivityCalculation.R

```diff
@@ -0,0 +1,112 @@
+#' simulation <- loadSimulation(simPath)
+#' outputPaths <- "Organism|PeripheralVenousBlood|Aciclovir|Plasma (Peripheral Venous Blood)"
+#' parameterPaths <- c(
+#'   "Aciclovir|Lipophilicity",
+#'   "Applications|IV 250mg 10min|Application_1|ProtocolSchemaItem|Dose",
+#'   "Neighborhoods|Kidney_pls_Kidney_ur|Aciclovir|Glomerular Filtration-GFR|GFR fraction"
+#' )
+#'
+#' # extract the results into a list of dataframes
+#' sensitivityCalculation(
+#'   simulation = simulation,
+#'   outputPaths = outputPaths,
+#'   parameterPaths = parameterPaths
+#' )
+#' @export
+sensitivityCalculation <- function(simulation,
+                                   outputPaths,
+                                   parameterPaths,
+                                   variationRange = c(seq(0.1, 1, by = 0.1), seq(2, 10, by = 1)),
+                                   pkParameters = c("C_max", "t_max", "AUC_inf"),
+                                   pkDataFilePath = NULL) {
+  # fail fast if specified PK parameters are not part of standard `enum`
+  validateIsIncluded(pkParameters, names(ospsuite::StandardPKParameter))
+
+  # set outputs to the provided path
+  clearOutputs(simulation)
+  addOutputs(outputPaths, simulation)
+
+  # save raw names of paths to create named list
+  parameterPathNames <- parameterPaths
+
+  # create simulation batch for efficient calculations
+  parameterPaths <- getAllParametersMatching(parameterPaths, simulation)
+
+  # extract dataframes with results
+  batchResults <- purrr::map(
+    .x = parameterPaths,
+    .f = ~ .extractSimBatchResults(
+      simulation,
+      parameterPath = .x,
+      variationRange = variationRange
+    )
+  )
+
+  # name list with name of each parameter path
+  batchResults <- purrr::set_names(batchResults, parameterPathNames)
+
+  # extract dataframe PK parameters
+  pkData <- .simResultsToPKDataFrame(batchResults, parameterPaths)
+
+  # filter out unneeded PK parameters
+  if (!is.null(pkParameters)) {
+    pkData <- dplyr::filter(pkData, PKParameter %in% pkParameters)
+  }
+
+  # add a new `.rowid` column with unique name for each output path
+  pkData <- .addRowid(pkData)
+
+  # write each wide dataframe in a list to a separate sheet in Excel
+  if (!is.null(pkDataFilePath)) {
+    # convert tidy data to wide format and add them to a list
+    pkData_wide_list <- purrr::map(
+      .x = pkData %>% split(.$.rowid),
+      .f = ~ .convertToWide(.x)
+    )
+
+    writexl::write_xlsx(pkData_wide_list, pkDataFilePath)
+  }
+
+  # remove the unneeded column
+  pkData <- dplyr::select(pkData, -c(".rowid"))
+
+  # final list with needed objects and dataframes for plotting functions
+  results <- list(
+    "SimulationResults" = batchResults,
+    "outputPaths" = outputPaths,
+    "parameterPaths" = parameterPaths,
+    "pkData" = pkData
+  )
+
+  # add additional attribute, which will be helpful for plotting methods to
+  # recognize this object
+  class(results) <- c("SensitivityAnalysis", class(results))
```

Comment:
I am not sure what it does, but `ospsuite` already implements a R6 class `SensitivityAnalysis`. Could this be a problem?

Reply:
Oh?

Yes, this can create problems. I will rename it `SensitivityCalculation`.

Reply:
That way, there will also be consistency between the name of the function creating the object and the object itself.